quiche/src/lib.rs

```diff
@@ -7961,13 +8134,143 @@ impl TransportParams {
+        let stateless_reset_token = u128::from_be_bytes(
+            b.get_bytes(16)?
+                .to_vec()
+                .try_into()
+                .unwrap(), // Since a 128 bit vector has just successfully been created above, this should never panic.
+        );
+
+        Ok(PreferredAddressParams::new(
+            addr_v4,
+            addr_v6,
+            cid,
+            stateless_reset_token,
+        ))
+    }
+
+    fn decode_v4(b: &mut octets::Octets) -> Result<Option<SocketAddrV4>> {
+        let mut addr_v4 = None;
+        let ip_v4_b = <[u8; 4]>::try_from(b.get_bytes(4)?.slice(4)?)
+            .map_err(|_| Error::InvalidTransportParam)?;
+        let ip_v4 = Ipv4Addr::from(ip_v4_b);
+
+        let port_v4 = b.get_u16()?;
+
+        if ip_v4.is_unspecified() ^ (port_v4 == 0) {
+            return Err(Error::InvalidTransportParam);
+        }
+
+        if !ip_v4.is_unspecified() {
+            addr_v4 = Some(SocketAddrV4::new(Ipv4Addr::from(ip_v4), port_v4));
+        }
+
+        Ok(addr_v4)
+    }
+
+    fn decode_v6(b: &mut octets::Octets) -> Result<Option<SocketAddrV6>> {
+        let mut addr_v6 = None;
+        let ip_v6_b = <[u8; 16]>::try_from(b.get_bytes(16)?.slice(16)?)
+            .map_err(|_| Error::InvalidTransportParam)?;
+        let ip_v6 = Ipv6Addr::from(ip_v6_b);
+        let port_v6 = b.get_u16()?;
+
+        if ip_v6.is_unspecified() ^ (port_v6 == 0) {
+            return Err(Error::InvalidTransportParam);
+        }
+
+        if !ip_v6.is_unspecified() {
+            addr_v6 = Some(SocketAddrV6::new(ip_v6, port_v6, 0, 0));
+        }
+
+        Ok(addr_v6)
+    }
+
+
+    /// Encode the preferred address parameters into a buffer.
+    ///
+    /// # Arguments
+    ///
+    /// * `preferred_address_params` - A reference to the preferred address parameters to be encoded.
+    /// * `buffer` - A reference to a buffer to encode the `preferred_address_params` into.
+    fn encode(
+        preferred_address_params: &PreferredAddressParams, buffer: &mut [u8],
+    ) -> Result<usize> {
+        let mut b = octets::OctetsMut::with_slice(buffer);
+
+        let (ip_v4, port_v4) = match preferred_address_params.addr_v4 {
+            Some(ip) => (ip.ip().octets(), ip.port()),
+            None => ([0; 4], 0),
+        };
+
+        b.put_bytes(&ip_v4)?;
+        b.put_u16(port_v4)?;
+
+        let (ip_v6, port_v6) = match preferred_address_params.addr_v6 {
+            Some(ip) => (ip.ip().octets(), ip.port()),
+            None => ([0; 16], 0),
+        };
+
+        b.put_bytes(&ip_v6)?;
+        b.put_u16(port_v6)?;
+
+        b.put_varint(preferred_address_params.connection_id.len() as u64)?;
+        b.put_bytes(&preferred_address_params.connection_id)?;
+
+        b.put_bytes(&preferred_address_params.stateless_reset_token.to_be_bytes())?;
+
+        Ok(b.off())
+    }
+
+    fn address_v4(&self) -> Option<(SocketAddrV4, ConnectionId, u128)> {
```

Comment:
Related to earlier comments, I'm not sure the of the value of these methods that wrap access to the fields in the parameter.